File: graph/mst/mst_kruskal.py
import sys

# Fast I/O
input = sys.stdin.readline

class DSU:
    def __init__(self, n):
        self.parent = list(range(n + 1))
        self.size = [1] * (n + 1)
    
    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    
    def union(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            if self.size[root_i] < self.size[root_j]:
                root_i, root_j = root_j, root_i
            self.parent[root_j] = root_i
            self.size[root_i] += self.size[root_j]
            return True
        return False

class KruskalMVP:
    """
    Kruskal's Algorithm for Minimum Spanning Tree (MST)
    Time Complexity: O(E log E) or O(E log V)
    """
    def __init__(self, n, edges):
        self.n = n
        self.edges = edges # List of (u, v, w)
    
    def solve(self):
        # Sort edges by weight
        self.edges.sort(key=lambda x: x[2])
        
        dsu = DSU(self.n)
        mst_weight = 0
        mst_edges = []
        edges_count = 0
        
        for u, v, w in self.edges:
            if dsu.union(u, v):
                mst_weight += w
                mst_edges.append((u, v, w))
                edges_count += 1
        
        if edges_count == self.n - 1:
            return mst_weight, mst_edges
        else:
            return -1, [] # Graph not connected

def solve():
    try:
        line1 = input().split()
        if not line1: return
        n, m = map(int, line1)
        
        edges = []
        for _ in range(m):
            u, v, w = map(int, input().split())
            edges.append((u, v, w))
            
        algo = KruskalMVP(n, edges)
        weight, mst = algo.solve()
        
        if weight != -1:
            print(f"MST Weight: {weight}")
            # for u, v, w in mst:
            #     print(f"{u} - {v}: {w}")
        else:
            print("IMPOSSIBLE")
            
    except ValueError:
        pass

if __name__ == "__main__":
    solve()

File: graph/mst/mst_kruskal.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Kruskal's Algorithm for Minimum Spanning Tree (MST)
 * Time Complexity: O(E log E)
 */

struct Edge {
    int u, v, w;
    bool operator<(const Edge& other) const {
        return w < other.w;
    }
};

struct DSU {
    vector<int> parent;
    vector<int> size;
    
    DSU(int n) {
        parent.resize(n + 1);
        size.assign(n + 1, 1);
        iota(parent.begin(), parent.end(), 0);
    }
    
    int find(int i) {
        if (parent[i] == i) return i;
        return parent[i] = find(parent[i]);
    }
    
    bool unite(int i, int j) {
        int root_i = find(i);
        int root_j = find(j);
        if (root_i != root_j) {
            if (size[root_i] < size[root_j]) swap(root_i, root_j);
            parent[root_j] = root_i;
            size[root_i] += size[root_j];
            return true;
        }
        return false;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;
    
    vector<Edge> edges;
    for (int i = 0; i < m; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        edges.push_back({u, v, w});
    }
    
    sort(edges.begin(), edges.end());
    
    DSU dsu(n);
    long long mst_weight = 0;
    int edges_count = 0;
    vector<Edge> mst_edges;
    
    for (const auto& edge : edges) {
        if (dsu.unite(edge.u, edge.v)) {
            mst_weight += edge.w;
            mst_edges.push_back(edge);
            edges_count++;
        }
    }
    
    if (edges_count == n - 1) {
        cout << "MST Weight: " << mst_weight << "\n";
    } else {
        cout << "IMPOSSIBLE\n";
    }
    
    return 0;
}
